proposals/nullable-reference-types-specification.md

```diff
@@ -324,6 +324,51 @@ The nullability of the type inferred for `x` is always *nullable*.
+
+The determination of what bounds to add in the first phase are enhanced as follows:
+
+If an argument `Ei` has a reference type, the type `U` used for inference depends on the null state of `Ei` as well as its declared type:
+- If the declared type is a nonnullable reference type `U0` or a nullable reference type `U0?` then
+    - if the null state of `Ei` is "not null" then `U` is `U0`
+    - if the null state of `Ei` is "maybe null" then `U` is `U0?`
+- Otherwise if `Ei` has a declared type, `U` is that type
+- Otherwise if `Ei` is `null` then `U` is the special bound `null`
+- Otherwise if `Ei` is `default` then `U` is the special bound `default`
+- Otherwise no inference is made.
+
+## Exact, upper-bound and lower-bound inferences
+
+In inferences *from* the type `U` *to* the type `V`, if `V` is a nullable reference type `V0?`, then `V0` is used instead of `V` in the following clauses.
+- If `V` is one of the unfixed type variables, `U` is added as an exact, upper or lower bound as before
+- Otherwise, if `U` is `null` or `default`, no inference is made
+- Otherwise, if `U` is a nullable reference type `U0?`, then `U0` is used instead of `U` in the subsequent clauses.
+
+The essence is that nullability that pertains directly to one of the unfixed type variables is preserved into its bounds. For the inferences that recurse further into the source and target types, on the other hand, nullability is ignored. It may or may not match, but if it doesn't, a warning will be issued later if the overload is chosen and applied.
+
+## Fixing
+
+The spec currently does not do a good job of describing what happens when multiple bounds are identity convertible to each other, but are different. This may happen between `object` and `dynamic`, between tuple types that differ only in element names, between types constructed thereof and now also between `C` and `C?` for reference types.
+
+To handle this we add more phases to fixing, which is now:
+
+1. Gather all the types in all the bounds as candidates, removing `?` from all that are nullable reference types
+2. Eliminate candidates based on requirements of exact, lower and upper bounds (ignoring `null` and `default` bounds)
+3. Eliminate candidates that do not have an implicit conversion to all the other candidates
+4. If the remaining candidates do not all have identity conversions to one another, then type inference fails
+5. *Merge* the remaining candidates as described below
+6. If the resulting candidate is a reference type or a nonnullable value type and *all* of the exact bounds or *any* of the lower bounds are nullable value types, nullable reference types, `null` or `default`, then `?` is added to the resulting candidate, making it a nullable value type or reference type.
```

Comment:
This rule doesn't say how to infer nullability from the nullability of upper bounds.

Reply:
Presuming you intend the nullability of upper bounds to affect inference.